### PR TITLE
soc: nrf53: allow configuring oscillators from DT

### PR DESCRIPTION
**Pull request by gmarull**
Check commit messages.

### CODE REVIEW

**Review by nordicjm — DISMISSED** [review]

**Review by carlescufi — CHANGES_REQUESTED** [review]
Needs an entry in the migration guide

File: dts/bindings/clock/nordic,nrf53x-hfxo.yaml
Comment (masz-nordic) [inline dts/bindings/clock/nordic,nrf53x-hfxo.yaml:1]: Is separate binding (instead of using `nordic,nrf-hfxo.yaml`) needed in order to limit enum values?

File: dts/bindings/clock/nordic,nrf53x-oscillators.yaml
Comment (masz-nordic) [inline dts/bindings/clock/nordic,nrf53x-oscillators.yaml:1]: I think this is not the right path - with this approach we'd need to start introducing `nordic,nrf53x-uarte`, etc.
Presence of HW features could be described in properties of generic node instead?

File: dts/bindings/clock/nordic,nrf53x-hfxo.yaml
Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53x-hfxo.yaml:1]: yes, Zephyr bindings do not have advances capacilities like e.g. dt-schema where a single binding could be used

File: dts/bindings/clock/nordic,nrf53x-oscillators.yaml
Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53x-oscillators.yaml:1]: This follows more closely the DT-spec, where compatible implies one or another programming model. Zephyr is not consistent on this, though. Some vendors use the `v1`, `v2`, etc. scheme (ST/GD IIRC).

File: dts/bindings/clock/nordic,nrf53x-lfxo.yaml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+  load-capacitance-picofarad:
+    type: int
+    enum:
+      - 6
+      - 7
+      - 9
+    description: |
+      Load capacitance in picofarads. This property is only used when
+      load-capacitors is set to "internal".
```
[the remark this answers is not in the document]
Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53x-lfxo.yaml:30]: we have 3 possible states:

1. no load-capacitors specified: do nothing and rely on retained content
2. internal: use load-capacitance-picofarad
3. external: use external caps

with load-capacitance-picofarad alone, you can't determine the (1) case (present would be use internal, missing external, but 1st case?)

[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply dts/bindings/clock/nordic,nrf53x-lfxo.yaml:30]: I don't think the first state should be allowed :) Either you use internal caps, and if you do, you must specify their value, or you do not, at which case they are disconnected. Is sounds like option 1 is to simply not initialize the clock, that should not be allowed right?

[the remark this answers is not in the document]
Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53x-lfxo.yaml:30]: these are retained registers according to the spec.

[the remark this answers is not in the document]
Comment (bjarki-andreasen) [reply dts/bindings/clock/nordic,nrf53x-lfxo.yaml:30]: ok, so the idea is that a bootloader of TF-M will set these values, and zephyr will not update the value on init if` load-capacitors` is not specified. Am I wrong, and is this documented somewhere if not? :)

[the remark this answers is not in the document]
Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53x-lfxo.yaml:30]: I'm not sure about this use case TBH, I just tried to preserve the functionality :-)

**Comment by gmarull** [comment]
> Needs an entry in the migration guide

@carlescufi done

**Review by bjarki-andreasen — DISMISSED** [review]

**Review by fabiobaltieri — DISMISSED** [review]

File: soc/nordic/nrf53/soc.c [hunk 2]
```diff
@@ -490,7 +517,9 @@ static int nordicsemi_nrf53_init(void)
 	nrf_gpio_pin_control_select(PIN_XL2, NRF_GPIO_PIN_SEL_PERIPHERAL);
 #endif /* !defined(CONFIG_BUILD_WITH_TFM) */
 #endif /* defined(CONFIG_SOC_ENABLE_LFXO) */
-#if defined(CONFIG_SOC_HFXO_CAP_INTERNAL)
+#if defined(CONFIG_SOC_HFXO_CAP_INTERNAL) || \
+	(defined(NRF_SOC_SECURE_SUPPORTED) && \
```
Comment (anangl) [inline soc/nordic/nrf53/soc.c:521]:
```suggestion
	(defined(CONFIG_NRF_SOC_SECURE_SUPPORTED) && \
```

File: soc/nordic/nrf53/soc.c [hunk 3]
```diff
@@ -490,7 +517,9 @@ static int nordicsemi_nrf53_init(void)
...
+	 DT_ENUM_HAS_VALUE(DT_NODELABEL(hfxo), load_capacitors, internal))
```
Comment (anangl) [inline soc/nordic/nrf53/soc.c:522]:
```suggestion
	 DT_ENUM_HAS_VALUE(HXFO_NODE, load_capacitors, internal))
```

File: soc/nordic/nrf53/soc.c [hunk 4]
```diff
@@ -515,13 +544,15 @@ static int nordicsemi_nrf53_init(void)
...
 	 */
 	uint32_t capvalue =
-		((slope + 56) * (CONFIG_SOC_HFXO_CAP_INT_VALUE_X2 - 14)
+		((slope + 56) * (HFXO_CAP_VAL_X2 - 14)
 		 + ((offset - 8) << 4) + 32) >> 6;
 
 	nrf_oscillators_hfxo_cap_set(NRF_OSCILLATORS, true, capvalue);
-#elif defined(CONFIG_SOC_HFXO_CAP_EXTERNAL)
+#elif defined(CONFIG_SOC_HFXO_CAP_EXTERNAL) || \
+	DT_ENUM_HAS_VALUE(DT_NODELABEL(hfxo), load_capacitors, external)
```
Comment (anangl) [inline soc/nordic/nrf53/soc.c:552]:
```suggestion
	DT_ENUM_HAS_VALUE(HXFO_NODE, load_capacitors, external)
```

File: soc/nordic/nrf53/soc.c [hunk 5]
```diff
@@ -472,15 +503,11 @@ static int nordicsemi_nrf53_init(void)
...
-			NRF_OSCILLATORS_LFXO_CAP_6PF :
-		IS_ENABLED(CONFIG_SOC_LFXO_CAP_INT_7PF) ?
-			NRF_OSCILLATORS_LFXO_CAP_7PF :
-		IS_ENABLED(CONFIG_SOC_LFXO_CAP_INT_9PF) ?
-			NRF_OSCILLATORS_LFXO_CAP_9PF :
-			NRF_OSCILLATORS_LFXO_CAP_EXTERNAL);
+#ifdef CONFIG_SOC_NRF5340_CPUAPP
+#if defined(CONFIG_SOC_ENABLE_LFXO) || \
+	DT_ENUM_HAS_VALUE(LFXO_NODE, load_capacitors, external) || \
+	DT_ENUM_HAS_VALUE(LFXO_NODE, load_capacitors, internal)
```
Comment (anangl) [inline soc/nordic/nrf53/soc.c:509]: Why not just `#if defined(LFXO_CAP)`?

File: soc/nordic/nrf53/soc.c (hunk 3 above)
Comment (anangl) [inline soc/nordic/nrf53/soc.c:522]: Also here `#if defined(HFXO_CAP_VAL_X2)` could be used (`HFXO_CAP_VAL_X2` could be defined only when `CONFIG_NRF_SOC_SECURE_SUPPORTED`).

File: soc/nordic/nrf53/Kconfig [hunk 6]
```diff
@@ -246,10 +283,25 @@ config SOC_HFXO_CAP_DEFAULT
 
 config SOC_HFXO_CAP_EXTERNAL
 	bool "Use external load capacitors"
+	select DEPRECATED
+	help
+	  This option is deprecated, use DT instead. Example configuration:
+
+	  &lfxo {
```
Comment (anangl) [inline soc/nordic/nrf53/Kconfig:290]: `&hfxo`

File: soc/nordic/nrf53/Kconfig [hunk 7]
```diff
@@ -246,10 +283,25 @@ config SOC_HFXO_CAP_DEFAULT
...
+	  };
 
 config SOC_HFXO_CAP_INTERNAL
 	bool "Use internal load capacitors"
 	depends on NRF_SOC_SECURE_SUPPORTED
+	select DEPRECATED
+	help
+	  This option is deprecated, use DT instead. Example configuration:
+
+	  &lfxo {
```
Comment (anangl) [inline soc/nordic/nrf53/Kconfig:301]: `&hfxo`

File: soc/nordic/nrf53/Kconfig [hunk 8]
```diff
@@ -258,6 +310,13 @@ config SOC_HFXO_CAP_INT_VALUE_X2
 	depends on SOC_HFXO_CAP_INTERNAL
 	range 14 40
 	help
+	  This option is deprecated, use DT instead. Example configuration:
+
+	  &lfxo {
```
Comment (anangl) [inline soc/nordic/nrf53/Kconfig:315]: `&hfxo`

File: soc/nordic/nrf53/Kconfig [hunk 9]
```diff
@@ -246,10 +283,25 @@ config SOC_HFXO_CAP_DEFAULT
...
 	bool "Use internal load capacitors"
 	depends on NRF_SOC_SECURE_SUPPORTED
+	select DEPRECATED
+	help
+	  This option is deprecated, use DT instead. Example configuration:
+
+	  &lfxo {
+	      load-capacitors = "internal";
+	      ...
+	  };
```
Comment (anangl) [inline soc/nordic/nrf53/Kconfig:304]: Maybe it would be better to refer to the SOC_HFXO_CAP_INT_VALUE_X2 option regarding the example configuration instead?

**Review by anangl — CHANGES_REQUESTED** [review]

File: soc/nordic/nrf53/soc.c (hunk 2 above)
Comment (gmarull) [reply soc/nordic/nrf53/soc.c]: fixed

File: soc/nordic/nrf53/soc.c (hunk 4 above)
Comment (gmarull) [reply soc/nordic/nrf53/soc.c]: changed

File: soc/nordic/nrf53/soc.c (hunk 3 above)
Comment (gmarull) [reply soc/nordic/nrf53/soc.c]: thanks for the optimization, updated

File: soc/nordic/nrf53/Kconfig (hunk 6 above)
Comment (gmarull) [reply soc/nordic/nrf53/Kconfig]: fixed

File: soc/nordic/nrf53/Kconfig (hunk 8 above)
Comment (gmarull) [reply soc/nordic/nrf53/Kconfig]: fixed

File: soc/nordic/nrf53/Kconfig (hunk 7 above)
Comment (gmarull) [reply soc/nordic/nrf53/Kconfig]: fixed

File: soc/nordic/nrf53/Kconfig (hunk 9 above)
Comment (gmarull) [reply soc/nordic/nrf53/Kconfig]: added a comment

**Comment by fabiobaltieri** [comment]
@gmarull needs a rebase as well

**Comment by gmarull** [comment]
> @gmarull needs a rebase as well

done

**Review by fabiobaltieri — DISMISSED** [review]

File: soc/nordic/nrf53/Kconfig [hunk 10]
```diff
@@ -215,19 +219,49 @@ config SOC_ENABLE_LFXO
 choice SOC_LFXO_LOAD_CAPACITANCE
 	prompt "LFXO load capacitance"
 	depends on SOC_ENABLE_LFXO
-	default SOC_LFXO_CAP_INT_7PF
```
Comment (anangl) [inline soc/nordic/nrf53/Kconfig:218]: Why is this removed? Behavior of these deprecated options should not be changed.

File: dts/bindings/clock/nordic,nrf53x-lfxo.yaml [hunk 11]
```diff
@@ -0,0 +1,30 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: Nordic nRF low-frequency crystal oscillator (nRF53X series)
+
+compatible: "nordic,nrf53x-lfxo"
```
Comment (anangl) [inline dts/bindings/clock/nordic,nrf53x-lfxo.yaml:6]: It's nRF53 Series, not nRF53X, so I think the compatibles should be "nordic,nrf53-*". Actually, we already have "nordic,nrf53-flash-controller" in tree.

File: soc/nordic/nrf53/Kconfig (hunk 10 above)
Comment (gmarull) [reply soc/nordic/nrf53/Kconfig:218]: likely a mistake, re-added

File: dts/bindings/clock/nordic,nrf53x-lfxo.yaml (hunk 11 above)
Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53x-lfxo.yaml:6]: we actually have `SOC_SERIES_NRF53X`, also to be consistent with 52 where we have 52x.

Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53x-lfxo.yaml:6]: So please let me know what you prefer, I have no strong opinion here.

Comment (nordicjm) [reply dts/bindings/clock/nordic,nrf53x-lfxo.yaml:6]: For dts I'm not bothered either way, the problem is with the Kconfig symbol. Can see here the series is nrf53 https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nordic/soc.yml#L16 which means https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nordic/nrf53/Kconfig.soc#L6 should be SOC_SERIES_NRF53 not SOC_SERIES_NRF53X, this will completely break when the symbols are automatically generated by the build system, it hasn't been changed yet because that symbol is used in a lot of places, including the HAL, so that will need a big update. I would lean on nrf53 for dts, since nrf53x is an issue with the Kconfig symbol

Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53x-lfxo.yaml]: updated to `nrf53`

File: dts/bindings/clock/nordic,nrf53x-oscillators.yaml
Comment (anangl) [inline dts/bindings/clock/nordic,nrf53x-oscillators.yaml:1]: Name of this file should be corrected.

**Review by anangl — DISMISSED** [review]

File: dts/bindings/clock/nordic,nrf53x-oscillators.yaml
Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53x-oscillators.yaml:1]: oops, done

**Review by anangl — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

File: dts/bindings/clock/nordic,nrf53-hfxo.yaml [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+description: Nordic nRF high-frequency crystal oscillator (nRF53 series)
+
+compatible: "nordic,nrf53-hfxo"
+
+include: fixed-clock.yaml
+
+properties:
+  clock-frequency:
```
Comment (nika-nordic) [inline dts/bindings/clock/nordic,nrf53-hfxo.yaml:11]: I think it would be beneficial to introduce common compatible like `nordic,nrf53-clock.yaml`, that would contain common `clock-frequency` and `load-capacitors` properties, to be utilized by both `nrf53-hfxo` & `nrf53-lfxo`.
`clock-frequency` could be extracted to even more generic `nordic-nrf5x-clock.yaml` . This way we would increase number of bindings, but reduce duplication across bindings

Comment (gmarull) [reply dts/bindings/clock/nordic,nrf53-hfxo.yaml:11]: `clock-frequency` is already common (comes from `fixed-clock`), so the only one that can be made common is `load-capacitors`.

**Review by nordicjm — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Review by rerickson1 — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Review by masz-nordic — APPROVED** [review]
